### ISSUE TITLE
Performance of asyncssh through VPN

### ISSUE THREAD

**Issue by maciej-kisiel** [ronf/asyncssh#86]
Hi!

I'm using asyncssh in my project and I discovered some weird behaviour when trying to connect to multiple servers through a VPN.

I wrote a simple script that tries connecting to some hosts and execute a simple 'ls /' command. With the first client the command is supposed to succeed and the second one is supposed to fail with permission denied (incorrect key).

Using openssh when testing in both local network and VPN it takes around 4s to complete the first host and ~0s to complete the second one.
Using asyncssh when testing in local network it takes 2s (!) and ~0s respectively. But when I try to connect to the host using a VPN it takes around 4-7s to complete the second host (which doesn't even connect) and 6-9s to complete the first one.

So, my question:

Is there any way the asyncssh is not working well with VPNs? There is also much variation in connection times, I'll post it at the end.

Bonus question: how it is faster that openssh on local network? :)

10 loops of the first client using asyncssh through VPN:

```
real	0m6.349s
real	0m6.103s
real	0m6.129s
real	0m9.934s
real	0m9.820s
real	0m6.254s
real	0m6.104s
real	0m5.849s
real	0m5.974s
real	0m6.504s
```

10 loops of the second client using asyncssh through VPN:

```
real	0m3.816s
real	0m3.801s
real	0m3.887s
real	0m3.901s
real	0m3.878s
real	0m7.245s
real	0m3.820s
real	0m4.121s
real	0m4.108s
real	0m4.296s
```

Both clients with asyncio.gather using asyncssh through VPN:

```
real	0m16.941s
real	0m16.626s
real	0m9.510s
real	0m16.851s
real	0m17.188s
real	0m9.825s
real	0m17.013s
real	0m16.945s
real	0m17.556s
real	0m10.638s
```

Thanks for help!

**Comment by ronf** [comment]
I'm not aware of anything that would cause asyncssh to run noticeably slower over VPN, unless the bandwidth was really limited or the network path was much slower for some reason.

As for the 4 seconds vs 2 seconds, one possibility is that AsyncSSH and the version of OpenSSH you are running are requesting different key exchange or cipher algorithms. By default, AsyncSSH will try to use curve25519-sha256 for key exchange and chacha20-poly1305 as a cipher/MAC. Newer version of OpenSSH should be similar, but if you are running an older version, it might be picking something else that's more expensive.

Another possibility is that the size of the SSH server and/or client keys being used could be larger on one of the two servers vs. the other. That could make a big difference in connect time.

In general, all of these times seem really high to me. When I test on my local network here, a single run (including the overhead to start Python and load up the code) takes about 0.5 seconds, and if I run 10 connections in a row from a single Python client, it only takes about 2 seconds for all 10 connections to connect, run a command, and close.

What kind of times do you get back if you "ping" the server over VPN, compared to the local network? Also, what version of AsyncSSH are you testing here? Would you mind posting the test code you are using (with any hostnames and passwords removed, of course)?

**Comment by maciej-kisiel** [comment]
Sorry, I forgot to mention. The device that I'm connecting to successfully is an embedded device which has limited resources so asymmetric crypto may take some time. Therefore I would look at the difference between local network vs. VPN (which is not visible when using openssh) rather than absolute times. The second device that is supposed to fail to connect is a fast computer.

I'm running asyncssh 1.9.0 with python 3.5.2. Also, my openssh is 7.2.

The pings are ~1ms on local network and 20-30ms through VPN.

The script I run:

```
import asyncssh
import asyncio

addrs = [vpn_ip_1, vpn_ip_2]

async def run(addr):
    async with asyncssh.connect(addr, username='root', known_hosts=None, client_keys=('key')) as conn:
        result = await conn.run('ls /')
        return result.stdout

tasks = [run(addr) for addr in addrs]
loop = asyncio.get_event_loop()
results = loop.run_until_complete(asyncio.gather(*tasks, return_exceptions=True))
```

It may as well be that something's wrong with the VPN but the fact that openssh works fine is an argument against it.

**Comment by ronf** [comment]
Thanks. Nothing jumps out at me about the test code. 

Can you capture the output of running "ssh -vvv" to each of the devices? That would let us see what algorithms are being chosen and compare that against what AsyncSSH is likely to select. Also, is the client key in 'key' the same as the client key OpenSSH is using in both cases?

One other thing that might be interesting is to do a packet capture of the two cases. While a lot of the connection will be encrypted data, we'll still be able to see the timing of when data is sent over the connection, and that might provide a clue about what parts of the process are taking time. We could then potentially add some debugging into AsyncSSH to refine that further.

**Comment by maciej-kisiel** [comment]
Yes, I also planned to check what's going on using wireshark. I'll get back to you with the results.

**Comment by maciej-kisiel** [comment]
Sorry for not writing for such a long time. After long debugging it turned out that incorrect interface MTU was causing this behaviour. It required a very specific set of circumstances to be present. I needed to be in one specific physical location, using asyncssh and VPN. If one of these conditions was not satisfied everything was working properly.

My hypothesis is that the additional VPN headers caused fragmentation of some big packets that were necessary in the process of setting up the SSH connection. Apparently, the ISP in that physical location was not handling such packets correctly which caused the delay, but it's my guess. Openssh was working better because the communication was slightly different is its case (it gave a different set of available encryption algorithms and hmacs) and it didn't contain such a big packet as in asyncssh's exchange.

After lowering the MTU from 1500 to 1300 everything works fine and asyncssh is even faster than openssh. This issue may be closed, I hope that it will save someone the hours I spent to debug this.

Thank you for your help and sorry for any inconvenience.
